src/MonoMod.Core/Platforms/Runtimes/FxCoreBaseRuntime.cs

```diff
@@ -68,8 +68,12 @@ TypeCode.Int32 or TypeCode.UInt32 ||
 
             // all others are passed on stack, or in a return buffer
             if (isReturn)
-            {
-                return TypeClassification.ByReference;
+            {
+                var typeSize = type.GetManagedSize();
+                if (typeSize is 1 or 2 or 4)
+                    return TypeClassification.InRegister;
+                else
+                    return TypeClassification.ByReference;

```

Comment:
I can't say I'm super happy with this, as is. I think at some point we should go and inspect the ABI logic for x86 and redo it from that. That's for another time though I think.

Reply:
Could this be the place @nike4613 ?
https://github.com/dotnet/runtime/blob/29acb2fd71a363053f1fbe67dc4bc06c555f7b94/src/coreclr/jit/compiler.cpp#L906

Reply:
It looks like it, yes.